src/strings.js

```diff
@@ -40,54 +40,118 @@ function rewritePatternForDotAll(pattern) {
   return cachedRegExp[pattern];
 }
 
-engine.indexOf = function(coll, substr){
+engine.indexOf = function (coll, substr) {
   const str = misc.singleton(coll, 'String');
   return util.isEmpty(substr) || util.isEmpty(str) ? [] : str.indexOf(substr);
 };
 
-engine.substring = function(coll, start, length){
+engine.substring = function (coll, start, length) {
   const str = misc.singleton(coll, 'String');
   if (util.isEmpty(str) || util.isEmpty(start) || start < 0 || start >= str.length) {
-    return  [];
+    return [];
   }
   if (length === undefined || util.isEmpty(length)) {
     return str.substring(start);
   }
   return str.substring(start, start + length);
 };
 
-engine.startsWith = function(coll, prefix){
+engine.startsWith = function (coll, prefix) {
   const str = misc.singleton(coll, 'String');
   return util.isEmpty(prefix) || util.isEmpty(str) ? [] : str.startsWith(prefix);
 };
 
-engine.endsWith = function(coll, postfix) {
+engine.endsWith = function (coll, postfix) {
   const str = misc.singleton(coll, 'String');
   return util.isEmpty(postfix) || util.isEmpty(str) ? [] : str.endsWith(postfix);
 };
 
-engine.containsFn = function(coll, substr){
+engine.containsFn = function (coll, substr) {
   const str = misc.singleton(coll, 'String');
   return util.isEmpty(substr) || util.isEmpty(str) ? [] : str.includes(substr);
 };
 
-engine.upper = function(coll){
+engine.upper = function (coll) {
   const str = misc.singleton(coll, 'String');
   return util.isEmpty(str) ? [] : str.toUpperCase();
 };
 
-
-engine.lower = function(coll){
+engine.lower = function (coll) {
   const str = misc.singleton(coll, 'String');
   return util.isEmpty(str) ? [] : str.toLowerCase();
 };
 
+engine.joinFn = function (coll, separator) {
+  const stringValues = coll.map((n) => {
+    const d = util.valData(n);
+    if (typeof d === "string") {
+      return d;
+    }
+    throw new Error('Join requires a collection of strings.');
+  });
+  if (separator === undefined) {
+    separator = "";
+  }
+  return stringValues.join(separator);
+};
+
+engine.splitFn = function (coll, separator) {
+  const strToSplit = misc.singleton(coll, 'String');
+  return util.isEmpty(strToSplit) ? [] : strToSplit.split(separator);
+};
+
+engine.trimFn = function (coll) {
+  const strToTrim = misc.singleton(coll, 'String');
+  return util.isEmpty(strToTrim) ? [] : strToTrim.trim();
+};
+
+// encoding/decoding
+engine.encodeFn = function (coll, format) {
+  const strToEncode = misc.singleton(coll, 'String');
+  if (util.isEmpty(strToEncode)){
+    return [];
+  }
+  if (format === 'urlbase64' || format === 'base64url'){
+    return btoa(strToEncode).replace(/\+/g, '-').replace(/\//g, '_');
+  }
+  if (format === 'base64'){
+    return btoa(strToEncode);
+  }
+  if (format === 'hex'){
+    return  Array.from(strToEncode).map(c => 
+      c.charCodeAt(0) < 128 ? c.charCodeAt(0).toString(16) : 
+      encodeURIComponent(c).replace(/\%/g,'')
```

Comment:
ESLint errors (see `npm test` output):
  123:1   error  Expected indentation of 8 spaces but found 6  indent
  123:38  error  Unnecessary escape character: \%              no-useless-escape